[python-package] Fix inconsistency in `predict()` output shape for 1-tree models

RektPunk opened the pull request:
Fixes https://github.com/microsoft/LightGBM/issues/6737

StrikerRUS commented on python-package/lightgbm/basic.py:
```diff
@@ -1248,7 +1248,7 @@ def predict(
             preds = preds.astype(np.int32)
         is_sparse = isinstance(preds, (list, scipy.sparse.spmatrix))
-        if not is_sparse and preds.size != nrow:
+        if not is_sparse and preds.size != nrow or pred_leaf:
```
Should we add `or pred_contrib` here as well?..

> for normal and raw score, its length is equal to `num_class * num_data`;
for leaf index, its length is equal to `num_class * num_data * num_iteration`;
for feature contributions, its length is equal to `num_class * num_data * (num_feature + 1)`.
https://lightgbm.readthedocs.io/en/latest/C-API.html#c.LGBM_BoosterPredictForMat

RektPunk replied:
I think it seems unnecessary since there's no place where the shape is used after `pred_contrib=True`. `pred_leaf` is necessary because of https://github.com/microsoft/LightGBM/blob/b33a12ea3883f306388e69f12ceb421b1ee7ec29/python-package/lightgbm/basic.py#L4882-L4888.

jameslamb replied:
> there's no place where the shape is used after pred_contrib=True.

@RektPunk `Booster.predict()` is an important part of LightGBM's public API. If it's producing incorrect results, that needs to be fixed.

We've set the expectation that `.predict(X, pred_leaf=True)`, for example, returns a matrix with shape `(X.shape[0], num_trees)` for regression, and right now that is not true for a single-tree model.

```python
import lightgbm as lgb
from sklearn.datasets import make_regression

X, y = make_regression(n_samples=10_000, n_features=10)

dtrain = lgb.Dataset(X, label=y)
params = {
    "objective": "regression",
    "verbosity": -1,
}

lgb.train(params, dtrain, num_boost_round=1).predict(X, pred_leaf=True).shape
# (10000,)

lgb.train(params, dtrain, num_boost_round=2).predict(X, pred_leaf=True).shape
# (10000, 2)
```

Even if no other code inside `lightgbm` referenced the `.shape` attribute of that output, this would still be a bug and we'd still want to fix it... because any other code using `lightgbm` should be able to form a consistent expectation of the return shape, and not need to have a special case for 1-tree models.

> *Should we add `or pred_contrib` here as well?*

I think we should, for completeness, but the existing condition is technically already enough to get the correct output shape.

`preds.size != nrow` will always be true for `pred_contrib`, because the output contains 1 value per feature AND the Shapley base values.

In other words... it will always have `nrow*(num_features + 1)` elements, so the reshaping will always be done, so there's no bug here for `pred_contrib` like there is for `pred_leaf`.

So please, let's do the following:

1. change this condition to `if not is_sparse and preds.size != nrow or (pred_leaf or pred_contrib)`
2. add explicit tests checking the output shapes for `pred_leaf` and `pred_contrib` with 1-iteration and 2-iteration models (using the snippet I provided above as a reference)

@RektPunk if you'd prefer that I add those additional tests, let me know and I'll push to your branch.

RektPunk replied:
Thanks for quick review. As you said, the condition is changed in [088a2b9](https://github.com/microsoft/LightGBM/pull/6753/commits/088a2b99e5c9dd78cf9ed27711cb0eec2c9dc00e),  and explicit test is added in [ba39a6f](https://github.com/microsoft/LightGBM/pull/6753/commits/ba39a6f3b466c7f0833751ee6cc300da6127d366). I'm not sure the exact location of newly added test though.

RektPunk replied:
I changed the condition as `if not is_sparse and (preds.size != nrow or pred_leaf or pred_contrib)` since previous one make test failed: [link](https://github.com/microsoft/LightGBM/actions/runs/12336081131/job/34428246842).

jameslamb replied:
Ah great, thank you! Sorry for suggesting the wrong form... I never remember exactly how a condition like `if A and B or C or D` is evaluated in Python 😅 

The parentheses make it easier to understand.

RektPunk replied:
No worries at all, Thank you so much for your kind suggestion :)

jameslamb reviewed (changes requested):
Thanks for this! I tested this locally and confirmed that this fixes the reported bug... thanks for diagnosing that and taking the time to submit a patch!

I left some suggestions on improving `lightgbm`'s tests to improve our confidence in this and prevent future refactorings from re-introducing bugs like this.

I've also changed the title of the PR, to reflect the fact that the bug here is not specific to `refit()`.

RektPunk commented on tests/python_package_test/test_engine.py:
```diff
@@ -2307,6 +2307,15 @@ def test_refit():
+    X, y = load_breast_cancer(return_X_y=True)
+    lgb_train = lgb.Dataset(X, label=y)
+    params = {"objective": "binary", "num_trees": 1, "verbosity": -1}
+    model = lgb.train(params, lgb_train, num_boost_round=1)
```
Sure, It is fixed in https://github.com/microsoft/LightGBM/pull/6753/commits/b6bb3c93da3c28dfbdaf11748f6b03015a991155, and also regression example is added in https://github.com/microsoft/LightGBM/pull/6753/commits/46a0ddc7d01402e4607471605e416e572822927a. For multi-class problem, since `preds.size != nrow`, I think there is no fail without this fix.

jameslamb replied:
> For multi-class problem, since preds.size != nrow, I think there is no fail without this fix.

That's fine, but please add a multi-class classification test anyway. That'll help prevent us from introducing this bug for multi-class classification in the future.

jameslamb commented on tests/python_package_test/test_engine.py:
```diff
@@ -2307,6 +2307,30 @@ def test_refit():
+    dtrain = lgb.Dataset(X, label=y)
+    params = {"objective": "regression", "verbosity": -1}
+    assert lgb.train(params, dtrain, num_boost_round=1).predict(X, pred_leaf=True).shape == (10_000, 1)
+    assert lgb.train(params, dtrain, num_boost_round=2).predict(X, pred_leaf=True).shape == (10_000, 2)
```
Thanks!

Please put this test down here by other `.predict()` tests:

https://github.com/microsoft/LightGBM/blob/b33a12ea3883f306388e69f12ceb421b1ee7ec29/tests/python_package_test/test_engine.py#L3827

And let's please:

* test all of `.predict(X, pred_leaf=True)`, `.predict(X, pred_contrib=True)`, and `.predict(X)`
* change the tests names to e.g. `test_predict_regression_output_shape`
* add similar tests for binary classification and 3-class multi-class classification

RektPunk replied:
Thanks! I add multiple shape check tests for `predict` including `pred_leaf` and `pred_contrib` for binary, multiclass test as you mentioned in https://github.com/microsoft/LightGBM/pull/6753/commits/d73f1894e7949e4cb7adb086d9d5c1288205cf1d

jameslamb commented on tests/python_package_test/test_engine.py:
```diff
@@ -2307,6 +2307,30 @@ def test_refit():
+    params = {"objective": "regression", "verbosity": -1}
+    model = lgb.train(params, lgb_train, num_boost_round=1)
+    model_refit = model.refit(X, y)
+    assert isinstance(model_refit, lgb.Booster)
```
If this test for regression is going to repeat all of the same code (totally fine, repetition in test code can be helpful!), then let's please just make it a separate test case.

```python
def test_refit_with_one_tree_regression():
   ...

def test_refit_with_one_tree_binary_classification():
   ...

def test_refit_with_one_tree_multiclass_classification():
   ...
```

That way, the test could be targeted individually like

```shell
pytest './tests/python_package_test/test_engine.py::test_refit_with_one_tree_binary_classification'
```

RektPunk replied:
Thanks, I add multiclass example and split the test into in https://github.com/microsoft/LightGBM/pull/6753/commits/03e41c61f919bf718d8c07fe416733e59e888611.

jameslamb reviewed (changes requested):
Thanks, left some more suggestions for the tests.

jameslamb commented on tests/python_package_test/test_engine.py:
```diff
@@ -15,7 +15,13 @@
+    make_blobs,
+    make_classification,
+    make_multilabel_classification,
+    make_regression,
```
```suggestion
```

Instead of introducing an import of `sklearn.datasets.make_regression`, please use the wrapper we have around it:

https://github.com/microsoft/LightGBM/blob/480600b3afaf2a0a6f32cf417edf9567f625b2c3/tests/python_package_test/utils.py#L118-L122

We like to do that because it makes the tests run faster... every invocation after the first one gets a cached dataset in memory instead of re-generating a new one.

RektPunk replied:
Thank you for kind review 👍 , I removed the `make_regression`.

jameslamb reviewed (changes requested):
Thanks for the tests! I left some small suggestions for re-organizing them.

Those are mostly minor personal preference, so if you'd prefer that I push such changes here directly let me know and I'd be happy to.

RektPunk commented on tests/python_package_test/test_engine.py:
```diff
@@ -2307,6 +2313,33 @@ def test_refit():
 
 
+def test_refit_with_one_tree_regression():
+    X, y = make_regression(n_samples=10_000, n_features=10)
```
Sure, I changed test with smaller example in https://github.com/microsoft/LightGBM/pull/6753/commits/6ad5c498c6bacd1919a83266d060c8e50a8af2b6

RektPunk commented on tests/python_package_test/test_engine.py:
```diff
@@ -3851,6 +3884,69 @@ def test_predict_stump(rng, use_init_score):
+    X, y = make_regression(n_samples=10_000, n_features=10)
+    dtrain = lgb.Dataset(X, label=y)
+    params = {"objective": "regression", "verbosity": -1}
+    assert lgb.train(params, dtrain, num_boost_round=1).predict(X).shape == (10_000,)
```
Thank you, the dupliucated tests are removes as you mentioned in https://github.com/microsoft/LightGBM/pull/6753/commits/2b4bfd72dfcecf1c218f303bcb3dc050244c011d.

jameslamb replied:
looks great, thanks!

RektPunk commented:
Hi @jameslamb, just a quick follow-up on the review. I’ve addressed the comments—please let me know if there’s anything else I should look into. Thanks!

jameslamb reviewed (approved):
This looks great, thanks very much @RektPunk ! We really appreciate you picking up and fixing this bug 😊
